### PR TITLE
CI: Add build matrix

### PR DESCRIPTION
**Pull request by mfkl**
- Refactor configuration properties that could be shared with Directory.Build.Props.
- Enforce a .NET SDK version with global.json for developers.
- Introduce build matrix to test on various .NET TFMs, runtimes and hosts (credits to @jgiannuzzi).

The publish step is the only one which I haven't tested, admittedly.

Just noticed that we have unit tests for Aspnet Core, I will add them to CI runs in a follow up PR (this one may already be large enough).

I've had intermittent test failures, but now I cannot reproduce... Will trigger several GitHub Action runs to try to repro again (never managed to repro locally).

### CODE REVIEW

**Comment by mfkl** [comment]
Ah well, I guess that PR run managed to repro it :D

```
[xUnit.net 00:00:01.41]     Consul.Test.SemaphoreTest.Semaphore_OneShot [FAIL]
[xUnit.net 00:00:01.41]       System.InvalidOperationException : Failed to create session
[xUnit.net 00:00:01.41]       ---- Consul.ConsulRequestException : Unexpected response, status code InternalServerError: Missing node registration
[xUnit.net 00:00:01.41]       Stack Trace:
[xUnit.net 00:00:01.41]         D:\a\consuldotnet\consuldotnet\Consul\Semaphore.cs(343,0): at Consul.Semaphore.Acquire(CancellationToken ct)
[xUnit.net 00:00:01.41]         D:\a\consuldotnet\consuldotnet\Consul\Semaphore.cs(456,0): at Consul.Semaphore.Acquire(CancellationToken ct)
[xUnit.net 00:00:01.41]         D:\a\consuldotnet\consuldotnet\Consul.Test\SemaphoreTest.cs(111,0): at Consul.Test.SemaphoreTest.Semaphore_OneShot()
[xUnit.net 00:00:01.41]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:00:01.41]         ----- Inner Stack Trace -----
[xUnit.net 00:00:01.41]         D:\a\consuldotnet\consuldotnet\Consul\Client_PutRequests.cs(356,0): at Consul.PutRequest`2.Execute(CancellationToken ct)
[xUnit.net 00:00:01.41]         D:\a\consuldotnet\consuldotnet\Consul\Session.cs(296,0): at Consul.Session.Create(SessionEntry se, WriteOptions q, CancellationToken ct)
[xUnit.net 00:00:01.41]         D:\a\consuldotnet\consuldotnet\Consul\Semaphore.cs(668,0): at Consul.Semaphore.CreateSession()
[xUnit.net 00:00:01.41]         D:\a\consuldotnet\consuldotnet\Consul\Semaphore.cs(337,0): at Consul.Semaphore.Acquire(CancellationToken ct)
Error: Consul.Test.SemaphoreTest.Semaphore_OneShot: System.InvalidOperationException : Failed to create session
  Failed Consul.Test.SemaphoreTest.Semaphore_OneShot [684 ms]
  Error Message:
   System.InvalidOperationException : Failed to create session
---- Consul.ConsulRequestException : Unexpected response, status code InternalServerError: Missing node registration
  Stack Trace:
     at Consul.Semaphore.Acquire(CancellationToken ct) in D:\a\consuldotnet\consuldotnet\Consul\Semaphore.cs:line 343
   at Consul.Semaphore.Acquire(CancellationToken ct) in D:\a\consuldotnet\consuldotnet\Consul\Semaphore.cs:line 456
   at Consul.Test.SemaphoreTest.Semaphore_OneShot() in D:\a\consuldotnet\consuldotnet\Consul.Test\SemaphoreTest.cs:line 111
--- End of stack trace from previous location where exception was thrown ---
----- Inner Stack Trace -----
   at Consul.PutRequest`2.Execute(CancellationToken ct) in D:\a\consuldotnet\consuldotnet\Consul\Client_PutRequests.cs:line 356
   at Consul.Session.Create(SessionEntry se, WriteOptions q, CancellationToken ct) in D:\a\consuldotnet\consuldotnet\Consul\Session.cs:line 296
   at Consul.Semaphore.CreateSession() in D:\a\consuldotnet\consuldotnet\Consul\Semaphore.cs:line 668
   at Consul.Semaphore.Acquire(CancellationToken ct) in D:\a\consuldotnet\consuldotnet\Consul\Semaphore.cs:line 337
```

Will investigate.

**Comment by mfkl** [comment]
That's another one, intermittent as well... (see [this successful run](https://github.com/mfkl/consuldotnet/actions/runs/859565906) a few minutes ago).

```
Run actions/setup-dotnet@v1
  with:
    dotnet-version: 2.1.x
  env:
    CONSUL_VERSION: 1.6.10
    DOTNET_SKIP_FIRST_TIME_EXPERIENCE: true
    DOTNET_CLI_TELEMETRY_OPTOUT: true
Error: connect ECONNREFUSED 52.241.88.84:443
```

**Comment by mfkl** [comment]
new failing tests are
```
[xUnit.net 00:00:01.55]     Consul.Test.LockTest.Lock_MultithreadedRelease [FAIL]
[xUnit.net 00:00:01.55]       System.Net.Http.HttpRequestException : No connection could be made because the target machine actively refused it. (127.0.0.1:8500)
[xUnit.net 00:00:01.55]       ---- System.Net.Sockets.SocketException : No connection could be made because the target machine actively refused it.
[xUnit.net 00:00:01.55]       Stack Trace:
[xUnit.net 00:00:01.55]            at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
[xUnit.net 00:00:01.55]            at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[xUnit.net 00:00:01.55]            at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[xUnit.net 00:00:01.55]            at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[xUnit.net 00:00:01.55]            at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
[xUnit.net 00:00:01.55]            at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[xUnit.net 00:00:01.55]            at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
[xUnit.net 00:00:01.55]         D:\a\consuldotnet\consuldotnet\Consul\Client_GetRequests.cs(78,0): at Consul.GetRequest`1.Execute(CancellationToken ct)
[xUnit.net 00:00:01.55]         D:\a\consuldotnet\consuldotnet\Consul\Status.cs(39,0): at Consul.Status.Leader(CancellationToken ct)
[xUnit.net 00:00:01.55]         D:\a\consuldotnet\consuldotnet\Consul.Test\BaseFixture.cs(36,0): at Consul.Test.BaseFixture.InitializeAsync()
[xUnit.net 00:00:01.55]         ----- Inner Stack Trace -----
[xUnit.net 00:00:01.55]            at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
[xUnit.net 00:00:01.55]            at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
[xUnit.net 00:00:01.55]            at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|283_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
[xUnit.net 00:00:01.55]            at System.Net.Http.HttpConnectionPool.DefaultConnectAsync(SocketsHttpConnectionContext context, CancellationToken cancellationToken)
[xUnit.net 00:00:01.55]            at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
Error: Consul.Test.LockTest.Lock_MultithreadedRelease: System.Net.Http.HttpRequestException : No connection could be made because the target machine actively refused it. (127.0.0.1:8500)
[xUnit.net 00:00:01.61]     Consul.Test.SemaphoreTest.Semaphore_OneShot [FAIL]
  Failed Consul.Test.LockTest.Lock_MultithreadedRelease [1 ms]
[xUnit.net 00:00:01.61]     Consul.Test.ACLReplicationTest.ACLReplication_GetStatus [FAIL]
  Error Message:
   System.Net.Http.HttpRequestException : No connection could be made because the target machine actively refused it. (127.0.0.1:8500)
---- System.Net.Sockets.SocketException : No connection could be made because the target machine actively refused it.
  Stack Trace:
     at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
   at Consul.GetRequest`1.Execute(CancellationToken ct) in D:\a\consuldotnet\consuldotnet\Consul\Client_GetRequests.cs:line 78
   at Consul.Status.Leader(CancellationToken ct) in D:\a\consuldotnet\consuldotnet\Consul\Status.cs:line 39
   at Consul.Test.BaseFixture.InitializeAsync() in D:\a\consuldotnet\consuldotnet\Consul.Test\BaseFixture.cs:line 36
----- Inner Stack Trace -----
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|283_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.DefaultConnectAsync(SocketsHttpConnectionContext context, CancellationToken cancellationToken)
   at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]       System.Net.Http.HttpRequestException : No connection could be made because the target machine actively refused it. (127.0.0.1:8500)
[xUnit.net 00:00:01.61]       ---- System.Net.Sockets.SocketException : No connection could be made because the target machine actively refused it.
[xUnit.net 00:00:01.61]       Stack Trace:
[xUnit.net 00:00:01.61]            at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]         D:\a\consuldotnet\consuldotnet\Consul\Client_GetRequests.cs(78,0): at Consul.GetRequest`1.Execute(CancellationToken ct)
[xUnit.net 00:00:01.61]         D:\a\consuldotnet\consuldotnet\Consul\Status.cs(39,0): at Consul.Status.Leader(CancellationToken ct)
[xUnit.net 00:00:01.61]         D:\a\consuldotnet\consuldotnet\Consul.Test\BaseFixture.cs(36,0): at Consul.Test.BaseFixture.InitializeAsync()
[xUnit.net 00:00:01.61]         ----- Inner Stack Trace -----
[xUnit.net 00:00:01.61]            at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
[xUnit.net 00:00:01.61]            at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|283_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.HttpConnectionPool.DefaultConnectAsync(SocketsHttpConnectionContext context, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]       System.Net.Http.HttpRequestException : No connection could be made because the target machine actively refused it. (127.0.0.1:8500)
[xUnit.net 00:00:01.61]       ---- System.Net.Sockets.SocketException : No connection could be made because the target machine actively refused it.
[xUnit.net 00:00:01.61]       Stack Trace:
[xUnit.net 00:00:01.61]            at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]         D:\a\consuldotnet\consuldotnet\Consul\Client_GetRequests.cs(78,0): at Consul.GetRequest`1.Execute(CancellationToken ct)
[xUnit.net 00:00:01.61]         D:\a\consuldotnet\consuldotnet\Consul\Status.cs(39,0): at Consul.Status.Leader(CancellationToken ct)
[xUnit.net 00:00:01.61]         D:\a\consuldotnet\consuldotnet\Consul.Test\BaseFixture.cs(36,0): at Consul.Test.BaseFixture.InitializeAsync()
[xUnit.net 00:00:01.61]         ----- Inner Stack Trace -----
[xUnit.net 00:00:01.61]            at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
[xUnit.net 00:00:01.61]            at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|283_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.HttpConnectionPool.DefaultConnectAsync(SocketsHttpConnectionContext context, CancellationToken cancellationToken)
[xUnit.net 00:00:01.61]            at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
Error: Consul.Test.SemaphoreTest.Semaphore_OneShot: System.Net.Http.HttpRequestException : No connection could be made because the target machine actively refused it. (127.0.0.1:8500)
  Failed Consul.Test.SemaphoreTest.Semaphore_OneShot [1 ms]
  Error Message:
   System.Net.Http.HttpRequestException : No connection could be made because the target machine actively refused it. (127.0.0.1:8500)
---- System.Net.Sockets.SocketException : No connection could be made because the target machine actively refused it.
  Stack Trace:
     at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
   at Consul.GetRequest`1.Execute(CancellationToken ct) in D:\a\consuldotnet\consuldotnet\Consul\Client_GetRequests.cs:line 78
   at Consul.Status.Leader(CancellationToken ct) in D:\a\consuldotnet\consuldotnet\Consul\Status.cs:line 39
   at Consul.Test.BaseFixture.InitializeAsync() in D:\a\consuldotnet\consuldotnet\Consul.Test\BaseFixture.cs:line 36
----- Inner Stack Trace -----
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|283_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.DefaultConnectAsync(SocketsHttpConnectionContext context, CancellationToken cancellationToken)
   at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
Error: Consul.Test.ACLReplicationTest.ACLReplication_GetStatus: System.Net.Http.HttpRequestException : No connection could be made because the target machine actively refused it. (127.0.0.1:8500)
  Failed Consul.Test.ACLReplicationTest.ACLReplication_GetStatus [1 ms]
  Error Message:
   System.Net.Http.HttpRequestException : No connection could be made because the target machine actively refused it. (127.0.0.1:8500)
---- System.Net.Sockets.SocketException : No connection could be made because the target machine actively refused it.
  Stack Trace:
     at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
   at Consul.GetRequest`1.Execute(CancellationToken ct) in D:\a\consuldotnet\consuldotnet\Consul\Client_GetRequests.cs:line 78
   at Consul.Status.Leader(CancellationToken ct) in D:\a\consuldotnet\consuldotnet\Consul\Status.cs:line 39
   at Consul.Test.BaseFixture.InitializeAsync() in D:\a\consuldotnet\consuldotnet\Consul.Test\BaseFixture.cs:line 36
----- Inner Stack Trace -----
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|283_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.DefaultConnectAsync(SocketsHttpConnectionContext context, CancellationToken cancellationToken)
   at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
```

File: Consul/Consul.csproj [hunk 1]
```diff
@@ -45,23 +25,21 @@
...
-    <Move SourceFiles="%(BuildArtifacts.RelativeDir)ILMerge\%(filename)%(extension)"
-          DestinationFiles="%(BuildArtifacts.RelativeDir)%(filename)%(extension)"
-          OverwriteReadOnlyFiles="true"/>
-    <Delete Files="@(ArtifactsToMerge)" />
-    <RemoveDir Directories="$(OutputPath)ILMerge" />
+    <MakeDir Directories="$(OutputPath)/ILRepack" />
+    <Exec WorkingDirectory="$(OutputPath)" Command="$(ILRepackCommand)" ContinueOnError="false" StandardOutputImportance="normal" />
+    <Move SourceFiles="$(OutputPath)ILRepack/$(AssemblyName).dll" DestinationFiles="$(OutputPath)$(AssemblyName).dll" OverwriteReadOnlyFiles="true" />
+    <RemoveDir Directories="$(OutputPath)ILRepack" />
+    <Message Importance="high" Text="Merged @(ArtifactsToMerge->'%(filename)%(extension)', ', ') into $(OutputPath)$(AssemblyName).dll" />
```
Comment (marcin-krystianc) [inline Consul/Consul.csproj:42]: Can we extract this change into separate PR ? I I assume that it is being done because `ILMerge` repository is now archived ?

File: Consul.Test/ACLReplicationTest.cs [hunk 2]
```diff
@@ -23,24 +23,8 @@
 
 namespace Consul.Test
 {
-    public class ACLReplicationTest : IDisposable
```
Comment (marcin-krystianc) [inline Consul.Test/ACLReplicationTest.cs:26]: Can we create separate PR for `*.cs` changes, it will make the matrix PR more coherent.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -6,20 +6,51 @@ on:
 
 env:
   CONSUL_VERSION: 1.6.10
```
Comment (marcin-krystianc) [inline .github/workflows/ci.yml:8]: No we are testing different version of consul, so I think that variable is not needed anymore

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -32,83 +63,102 @@ jobs:
...
           cd Consul.Test
           SYSTEM=$(uname | sed 's/MINGW.*/windows/' | tr A-Z a-z)
-          curl -sSL https://releases.hashicorp.com/consul/${CONSUL_VERSION}/consul_${CONSUL_VERSION}_${SYSTEM}_amd64.zip -o consul.zip
+          curl -sSL https://releases.hashicorp.com/consul/${{ matrix.consul }}/consul_${{ matrix.consul }}_${SYSTEM}_amd64.zip -o consul.zip
           unzip consul.zip
           rm consul.zip
-      - name: Build
-        run: dotnet build --configuration=Release -p:TreatWarningsAsErrors=true
+      - name: Build library
+        run: dotnet build --framework ${{ steps.dotnet.outputs.libframework }} --configuration=Release -p:Channel=${{ steps.dotnet.outputs.channel }} ${{ matrix.component }}
```
Comment (marcin-krystianc) [inline .github/workflows/ci.yml:113]: Do we actually use mono in the Ci or it is just a remnant of initial attempts ?

Comment (mfkl) [reply .github/workflows/ci.yml:113]: > Do we actually use mono in the Ci 

Doesn't look like it

File: Consul.Test/ACLReplicationTest.cs (hunk 2 above)
Comment (mfkl) [reply Consul.Test/ACLReplicationTest.cs:26]: Sure

File: Consul/Consul.csproj (hunk 1 above)
Comment (mfkl) [reply Consul/Consul.csproj:42]: Not sure, likely related to Mono https://github.com/jgiannuzzi/consuldotnet/commit/559e332d4374f3403de4f5fea6fc49a298e93426

File: .github/workflows/ci.yml (hunk 3 above)
Comment (mfkl) [reply .github/workflows/ci.yml]: Indeed, done

File: Consul.Test/ACLReplicationTest.cs (hunk 2 above)
Comment (mfkl) [reply Consul.Test/ACLReplicationTest.cs]: https://github.com/G-Research/consuldotnet/pull/88 done

**Comment by mfkl** [comment]
This PR build failure is expected until https://github.com/G-Research/consuldotnet/pull/88 is merged and this is rebased on top.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (marcin-krystianc) [reply .github/workflows/ci.yml:112]: So I think setting the `Channel` property is not needed.

File: Consul/Consul.csproj [hunk 5]
```diff
@@ -2,39 +2,19 @@
 
   <PropertyGroup>
     <VersionPrefix>1.6.10.1</VersionPrefix>
-    <TargetFrameworks Condition="'$(OS)' == 'Windows_NT'">netstandard2.0;net461</TargetFrameworks>
-    <TargetFrameworks Condition="'$(OS)' == 'Unix'">netstandard2.0</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' == 'Windows_NT' Or '$(Channel)' == 'mono'">$(TargetFrameworks);net461</TargetFrameworks>
```
Comment (marcin-krystianc) [inline Consul/Consul.csproj:6]: `dotnet5` sdk can build `net461` on any OS, so the is no need for this conditional definition of `TargetFramerworks`.

File: Consul.Test/Consul.Test.csproj [hunk 6]
```diff
@@ -30,7 +22,11 @@
     <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="1.0.2" />
   </ItemGroup>
 
-  <PropertyGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' OR '$(TargetFramework)' == 'netcoreapp3.1' ">
+  <ItemGroup Condition=" '$(TargetFramework)' != 'netcoreapp2.1' ">
+    <PackageReference Include="GitHubActionsTestLogger" Version="1.1.0" />
```
Comment (marcin-krystianc) [inline Consul.Test/Consul.Test.csproj:26]: `GitHubActionsTestLogger` works with 'netcoreapp2.1', just use version `1.2.0` and add the 
```
<CopyLocalLockFileAssemblies Condition="'$(TargetFramework)' == 'netcoreapp2.1'" >true</CopyLocalLockFileAssemblies>
```
to the test project as a workaround.

**Comment by mfkl** [comment]
As @marcin-krystianc told me privately, this PR is a bit too large to be reviewed comfortably. 

Several tightly related build improvements tagged along during the dev process and then dealing with the intermittent test failures added yet again new changes. We will take those changes as separate smaller PRs until the only diff here is the matrix YML change (and likely the separate packaging step, that depends on it).

**Comment by mfkl** [comment]
Superseded by https://github.com/G-Research/consuldotnet/pull/89
